### PR TITLE
Particle parser in manifest

### PR DESCRIPTION
**Pull request by mmandlis**

### CODE REVIEW

File: runtime/manifest-parser.peg [hunk 1]
```diff
@@ -38,36 +41,186 @@ ManifestItem
...
+      args,
+      affordance,
+      slots,
+      description
+    };
+  }
+
+ParticleItem
+  = ParticleInterface
+  / Affordance
```
Comment (dstockwell) [inline runtime/manifest-parser.peg]: ParticleAffordance?

File: runtime/manifest-parser.peg [hunk 2]
```diff
@@ -38,36 +41,186 @@ ManifestItem
...
+  / ParticleSlot
+  / ParticleDescription
+
+ParticleInterface
+  = name:upperIdent '(' args:ParticleArgumentList ')' eolWhiteSpace
```
Comment (dstockwell) [inline runtime/manifest-parser.peg]: TODO: make the argument list optional

(to fix that bug that requires at least one particle argument)

File: runtime/manifest-parser.peg [hunk 3]
```diff
@@ -38,36 +41,186 @@ ManifestItem
...
+  {
+    return { name, args }
```
Comment (dstockwell) [inline runtime/manifest-parser.peg]: nit semicolon and no whitespace needed for single line literal

`return {name, args};`

File: runtime/manifest-parser.peg [hunk 4]
```diff
@@ -38,36 +41,186 @@ ManifestItem
...
+  }
+
+ParticleArgumentList
+  = head:ParticleArgument tail:(',' whiteSpace ParticleArgument)*
+  {
+    return [head].concat(tail.map(a => a[2]));
+  }
+
+ParticleArgument
+  = direction:ParticleArgumentDirection whiteSpace tn:ParticleArgumentType whiteSpace name:(lowerIdent)
```
Comment (dstockwell) [inline runtime/manifest-parser.peg]: what is tn? can we just call this type?

Comment (dstockwell) [inline runtime/manifest-parser.peg]: no brackets needed around `lowerIdent` here

File: runtime/manifest-parser.peg [hunk 5]
```diff
@@ -38,36 +41,186 @@ ManifestItem
...
+  {
+    return {direction, type:tn, name};
+  }
+
+ParticleArgumentDirection
+  = 'in' / 'out' / 'inout'
+  {
+     // Do we still need ephemeral?
```
Comment (dstockwell) [inline runtime/manifest-parser.peg]: TODO?

File: runtime/manifest-parser.peg [hunk 6]
```diff
@@ -38,36 +41,186 @@ ManifestItem
...
+  }
+
+ParticleArgumentType
+  = (VariableType / PrimitiveType / ListType)
+ListType
+  = '[' pt:ParticleArgumentType ']' { return typeLiteral.viewOf(pt); }
+VariableType
+  = '~' name: VariableTypeName { return typeLiteral.namedTypeVariable(name); }
+VariableTypeName
+  = [a-z][a-zA-Z0-9]* { return text() }
```
Comment (dstockwell) [inline runtime/manifest-parser.peg]: is this the same as `lowerIdent`?

File: runtime/manifest-parser.peg [hunk 7]
```diff
@@ -38,36 +41,186 @@ ManifestItem
...
+PrimitiveType
+  = [A-Z][a-zA-Z0-9]* { return text() }
```
Comment (dstockwell) [inline runtime/manifest-parser.peg]: `upperIdent`?

File: runtime/manifest-parser.peg [hunk 8]
```diff
@@ -38,36 +41,186 @@ ManifestItem
...
+  = 'affordance' whiteSpace affordance:('dom-touch' / 'dom' / 'vr' / 'voice') eolWhiteSpace
+  { return {affordance}; }
+
+ParticleSlot
+  = isRequired:('must' whiteSpace)? 'consume' whiteSpace name:(lowerIdent) eolWhiteSpace
+    items:(Indent (SameIndent ParticleSlotItem)*)?
+  {
+    let formFactor = null;
+    let providedSlots = [];
+    items = items ? extractIndented(items) : [];
```
Comment (dstockwell) [inline runtime/manifest-parser.peg]: `items = optional(items, extractIndented, []);`

File: runtime/manifest-parser.peg [hunk 9]
```diff
@@ -38,36 +41,186 @@ ManifestItem
...
+    items.forEach(item => {
+      if (item.kind == 'provided-slot') {
+        providedSlots.push(item);
+      } else if (item.kind == 'form-factor') {
+        if (formFactor)
+          assert ('duplicate form factor for a slot');
```
Comment (dstockwell) [inline runtime/manifest-parser.peg]: Can you try to call `error('duplicate form...` here? That should cause the parser to throw it's regular error object with position information.

File: runtime/manifest.js [hunk 10]
```diff
@@ -68,9 +68,15 @@ class Manifest {
     registry = registry || {};
     position = position || {line: 1, column: 0};
 
-    let items = parser.parse(content);
+    let items = [];
+    try{
+      items = parser.parse(content);
+    } catch (e) {
+      console.log(e);
```
Comment (dstockwell) [inline runtime/manifest.js]: throw e; after console.log.

File: runtime/test/manifest-test.js [hunk 11]
```diff
@@ -38,14 +38,24 @@ describe('manifest', function() {
...
+          formFactor big
+          provide action
+            view list
+            formFactor big
+          provide preamble
+            formFactor medium
+          provide annotation
+        consume other
+        description
+          pattern \`hello world \${list}\`
```
Comment (dstockwell) [inline runtime/test/manifest-test.js]: is 'pattern' a special name? If yes, and there's just one, perhaps this should be on the same line as description:

```
  description `hello world ${list}`
    list `my special list`
```

File: runtime/manifest.js (hunk 10 above)
Comment (mmandlis) [inline runtime/manifest.js]: done

**Pull request merged**